### PR TITLE
Galois Python datastructures applications

### PR DESCRIPTION
**Pull request by arthurp**
This PR adds support for Galois data structures (InsertBag and LargeArray) to the Python and numba interfaces. This PR also includes an implementation of BFS using this new support.

This is very much a prototype. There is almost certainly a memory leak in the numba closure support and the method for supporting galois iteration over the data structures is not at all scalable.

### CODE REVIEW

**Comment by arthurp** [comment]
@gurbinder533, what do you think of this? I think we should view this as a discussion piece and we can decide what if any changes are needed to make this into a "real" implementation.

Would it be possible to make provide a dynamically polymorphic iterator interface for the galois loops? Like by providing a virtual implementation of the "blocked iterator" concept that the galois loops expect, and then wrapping the various other iterators in this dynamically dispatched implementation. It this is possible then we could dramatically simplify the iterable handling for the loops and make it somewhat extensible, so other modules could provide support for their data structures if needed.

File: python/galois/bfs_property_graph.py [hunk 1]
```diff
@@ -1,5 +1,92 @@
...
+from ._bfs_property_graph import bfs as cython_bfs, verify_bfs as cython_verify_bfs
+from .loops import do_all, do_all_operator
 from .property_graph import PropertyGraph
+from .timer import StatTimer
+
+
+@jit(nopython=True)
+def initialize(graph: PropertyGraph, source: int, distance: np.ndarray):
+    num_nodes = graph.num_nodes()
+    for n in range(num_nodes):
```
Comment (gurbinder533) [inline python/galois/bfs_property_graph.py:16]: Any reason for not using do_all here?

File: python/galois/bfs_property_graph.py [hunk 2]
```diff
@@ -1,5 +1,92 @@
...
+    if val < num_nodes:
+        max_dist.update(val)
+
+
+def verify_bfs(graph: PropertyGraph, source_i: int, property_id: int):
+    chunk_array = graph.get_node_property(property_id)
+    not_visited = GAccumulator[int](0)
+    max_dist = GReduceMax[int]()
+
+    do_all(range(len(chunk_array)),
```
Comment (gurbinder533) [inline python/galois/bfs_property_graph.py:42]: This is arrow::ChunkedArray, right? Shouldn't we loop over all the chunks?

File: python/galois/bfs_property_graph.py [hunk 3]
```diff
@@ -1,5 +1,92 @@
...
+        if distance[dst] == num_nodes:
+            distance[dst] = next_level
+            next.push(dst)
+
+
+def bfs_sync_pg(graph: PropertyGraph, source, property_name):
+    num_nodes = graph.num_nodes()
+    next_level = 0
+
+    curr = InsertBag['uint64_t']()
```
Comment (gurbinder533) [inline python/galois/bfs_property_graph.py:71]: Just to make sure, we can't pass arbitrary types to the InsertBag at the moment, right?

File: python/galois/bfs_property_graph.py [hunk 4]
```diff
@@ -1,5 +1,92 @@
...
+    while not next.empty():
+        curr.swap(next)
+        next.clear()
+        next_level += 1
+        do_all(curr,
+               bfs_sync_operator_pg(graph, next, next_level, distance),
+               steal=True, loop_name="bfs_sync_pg")
+    timer.stop()
+
+    graph.add_node_property(pyarrow.table({property_name: distance}))
```
Comment (gurbinder533) [inline python/galois/bfs_property_graph.py:88]: What is the behavior if the property "distance" already exists? Does it over-write?

File: python/galois/bfs_property_graph.py (hunk 3 above)
Comment (arthurp) [reply python/galois/bfs_property_graph.py:71]: No. And we will never be able to pass arbitrary types (unless we want to invoke the C++ compiler at runtime).

File: python/galois/bfs_property_graph.py (hunk 2 above)
Comment (arthurp) [reply python/galois/bfs_property_graph.py:42]: Currently I'm just taking the performance hit of indexing the chunked array using a single index (the generated code will iterate over the chunks to find the one with element I'm looking for). Down the road we can support iterating over the index set of chunked array and those indicies can be pairs (chunk_i, element_i) so the look ups will be fast. Or just iterate over the chunked array elements directly and hide the indexing entirely. I don't see any reason for the user to be aware of chunking, we should be able to hide it.

File: python/galois/bfs_property_graph.py (hunk 1 above)
Comment (arthurp) [reply python/galois/bfs_property_graph.py:16]: The Cython implementation didn't use do_all, I just paralleled that.

File: python/galois/bfs_property_graph.py (hunk 4 above)
Comment (arthurp) [reply python/galois/bfs_property_graph.py:88]: It raises `ValueError`. We could add an "update_node_property" or something like that that has replacement semantics if we want.

**Comment by arthurp** [comment]
BTW, I am __not very happy__ with the state of the code generation (the jinja stuff). If you see anything that is too opaque tell me and I'll try to fix it. there may be things that would actually be better __not__ to generate and instead just manage the duplication manually. I tend to make the mistake of over-using code generation.

File: python/galois/bfs_property_graph.py (hunk 1 above)
Comment (gurbinder533) [reply python/galois/bfs_property_graph.py:16]: We can use do_all here. Can be done later.

File: python/galois/bfs_property_graph.py (hunk 2 above)
Comment (gurbinder533) [reply python/galois/bfs_property_graph.py:42]: Yes, we should hide the indexing.

File: python/galois/_loops.pyx.jinja [hunk 5]
```diff
@@ -8,13 +8,27 @@ import cython
...
 import galois
 import galois.loops
 import galois.numba.closure
-from .cpp.libgalois cimport Galois
+
+from .cpp.libgalois cimport Galois, datastructures
+from .cpp.libgalois.graphs cimport Graph
+
+from .datastructures cimport InsertBag_uint64_t, LargeArray_uint64_t
+from .datastructures import InsertBag_uint64_t, LargeArray_uint64_t
```
Comment (gurbinder533) [inline python/galois/_loops.pyx.jinja:22]: Can we add InsertBag and LargeArray variant for uint32_t as well? Graph node ids are expected to be uin32_t at least in the current versions (v1) that we support.

Comment (arthurp) [reply python/galois/_loops.pyx.jinja:22]: We would have to change a lot more than just the InsertBag and LargeArray. The entire system currently assumes that node and edge ids are uint64_t. One thing we could do is have types that STORE uint32_t but who's API accepts and returns uint64_t. That might reduce the amount of code duplication.

**Pull request merged**